### ISSUE TITLE
How to constraint one of the coeffiencts

### ISSUE THREAD

**Issue by emindeniz** [CamDavidsonPilon/lifelines#371]
Hello,
I am interested in constraining one of the coefficients in my COX regression model. More specifically, I am building a survival model for breast cancer patients. One of the variables in the model is a binary variable indicating whether chemotherapy is given to the patients. However, calculating the effect of this variable from my data may be biased, instead I am interested in putting another constant for this variable calculated from randomized and well controlled clinical trials. What could be a quick hack for me to alter the source code for this?

Thanks for great work.

**Comment by CamDavidsonPilon** [comment]
Hey @emindeniz - can you describe what the constraint might look like? Ex: If it's a differentiable constraint, you could hack the code here: https://github.com/CamDavidsonPilon/lifelines/blob/master/lifelines/fitters/coxph_fitter.py#L209-L210

**Comment by emindeniz** [comment]
Hello,
What I mean by constraint is simply fixing the value to a constant. Perhaps briefly looking at statistical methods section of this paper might help:

https://breast-cancer-research.biomedcentral.com/articles/10.1186/s13058-017-0852-3

from this paper:

"The variables for the final models were selected by sequential backward elimination [14]. The effects of adjuvant chemotherapy and adjuvant hormone therapy were constrained to the effects reported for standard anthracycline-based chemotherapy and adjuvant tamoxifen from an updated analysis of the Early Breast Cancer Trialists Collaborative Group"

Basically, people know very accurately the hazard ratio for the effect of chemotherapy. And I want to use that value from the literature.

Thanks,

**Comment by CamDavidsonPilon** [comment]
A quick solution might be to edit the code I outlined (https://github.com/CamDavidsonPilon/lifelines/blob/master/lifelines/fitters/coxph_fitter.py#L209-L210)
into something like:
```python
                # let ix_constraint_col be the index of the column you wish to constrain. 
                g[ix_constraint_col] -= self.penalizer * (beta[ix_constraint_col] - -0.31)
                h[ix_constraint_col, ix_constraint_col] -= self.penalizer
```

and then have a very high penalizer in the initialisation:

```python
cph = CoxPHFitter(penalizer=10)
cph.fit(...)
```

This will _try_ to force that coefficient to be close to -0.31 (this code is untested though).

**Comment by emindeniz** [comment]
Great thanks! I guess the idea is to add the difference between the value calculated from the data and the constant back to the value at each step. I will give it a try.

**Comment by emindeniz** [comment]
I altered the code the following way. Interestingly if chemotherapy is the first column it shrinks its coefficient, making the other coefficients larger ( I guess that's expected). 

```
            if self.penalizer > 0:
                # add the gradient and hessian of the l2 term
                # let ix_constraint_col be the index of the column you wish to constrain. 
                ix_constraint_col = X.columns.get_loc("Chemotherapy")
                g[ix_constraint_col] -= self.penalizer * (beta[ix_constraint_col] - -0.31)
                h[ix_constraint_col, ix_constraint_col] -= self.penalizer
                
                #g -= self.penalizer * beta.T
                #h.flat[::d + 1] -= self.penalizer
```

However it only works if chemotherapy coefficient is the first column otherwise it produces the following error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-33-d09214779c41> in <module>()
      9 print(COX_dataset.shape)
     10 
---> 11 cph = COXPH_backward_elimination(COX_dataset,penalizer=100)
     12 final_factors = cph.summary.index.tolist()
     13 cph.print_summary()

<ipython-input-27-d445366e289f> in COXPH_backward_elimination(COX_dataset, penalizer)
      1 def COXPH_backward_elimination(COX_dataset,penalizer=0):
      2     cph = CoxPHFitter(penalizer=penalizer)
----> 3     cph.fit(COX_dataset , duration_col='Days', event_col='Vitality')
      4 
      5 

~/.conda/envs/compete/lib/python3.5/site-packages/lifelines/fitters/coxph_fitter.py in fit(self, df, duration_col, event_col, show_progress, initial_beta, include_likelihood, strata)
    311         hazards_ = self._newton_rhaphson(df, T, E, initial_beta=initial_beta,
    312                                          show_progress=show_progress,
--> 313                                          include_likelihood=include_likelihood)
    314 
    315         self.hazards_ = pd.DataFrame(hazards_.T, columns=df.columns, index=['coef']) / self._norm_std

~/.conda/envs/compete/lib/python3.5/site-packages/lifelines/fitters/coxph_fitter.py in _newton_rhaphson(self, X, T, E, initial_beta, step_size, precision, show_progress, include_likelihood)
    217                 # let ix_constraint_col be the index of the column you wish to constrain.
    218                 ix_constraint_col = X.columns.get_loc("Chemotherapy")
--> 219                 g[ix_constraint_col] -= self.penalizer * (beta[ix_constraint_col] - -0.31)
    220                 h[ix_constraint_col, ix_constraint_col] -= self.penalizer
    221 

IndexError: index 31 is out of bounds for axis 0 with size 1
```

**Comment by emindeniz** [comment]
Hello,
I have come up with this idea of a related workaround. What if I don't include chemotherapy variable at all during the calculation of other coefficients. Then in my final COX model I simply add the coefficient to the model (since I know the value of beta coefficient for chemotherapy variable anyway?). Would this be an acceptable solution?
Thanks,

**Comment by CamDavidsonPilon** [comment]
It doesn't sound very sound, unfortunately, unless all your other variables are totally uncorrelated with the chemotherapy variable (even then, I'm not so sure...). 

Can other Python or R packages handle constraining a coefficient? If so, can you show me?
